### PR TITLE
Introduce permit/forbid attributes matchers

### LINKED ISSUE #94: Proposal: alias or rename `permit_mass_assignment_of`
I'm bringing this up since it seems that there is some refactoring taking place with the attribute matcher(s).

Context: we're using Pundit with Graphiti (Graphiti is an api tooling gem that inserts a layer between models and controllers that is tailored to REST API concerns). The specifics of graphiti aren't relevant to the issue here, other than to know that there are use cases for Pundit policies to implement and define `permitted_attributes` that aren't strictly limited to use within controllers. Ergo, the verbiage of "mass assignment" doesn't really apply.

(For more context, the use of `permitted_attributes` is for providing per-attribute `reabable`/`writable` type control at the api resource layer: https://www.graphiti.dev/guides/concepts/resources#limiting-behavior)

## primary request

So the initial proposal is to consider renaming the attribute matchers as `permit_attributes`/`forbid_attributes` which I think still read well with the chained action matcher:

```
it { is_expected.to permit_attributes(%i[name description]) }
it { is_expected.to permit_attributes(%i[name description]).for_action(:read) }
```

The argument in favor of renaming is that `permit_attributes` has use _outside_ of controllers. In particular, Pundit itself makes no assumption since policies are POROs; they set an expectation that policies implement `permitted_attributes` and so pundit-matchers (being a matcher library for pundit) is really only concerned with testing policies and need not (should not, IMO) presume the context in which those policies and their methods are used. And since the method that the matcher is asserting against is quite literally `permitted_attributes`, it seems fitting that the matcher use that verbiage (permit/forbid _attributes).

## secondary request

Of course, rspec allows aliasing matchers. We've gone that route in the interim:

```
alias_matcher :permit_serialization_of, :permit_mass_assignment_of
alias_matcher :forbid_serialization_of, :forbid_mass_assignment_of
```

However, the downside is that the failure message (and test docstrings) descriptions are not overridden consistently. The docstrings generated from "permit" seem to take the value properly from aliased matcher name: `"is expected to permit serialization of ..."` but for some reason, the docstrings from the forbid alias is _not_ accurate: `"is expected to forbid mass assignment of  mass assignment of ..."`. I suspect this is because of how pundit-matchers is currently defining the forbid negated matcher, which will _hopefully_ be resolved by the refactor that is in progress. So to summarize the second request here (regardless if the matchers are renamed): the mechanism by which pundit-matchers defines both the permit and forbid matchers should use rspec's facilities in such a way that further aliases generate the correct docstrings.

### PR DESCRIPTION
**Pull request by tagliala**
Add `permit_attribute` and `permit_attributes` to match actions API

Also bumps version to 3.1.0

Close #94

### CODE REVIEW

**Comment by tagliala** [comment]
There is only a caveat, I don't know if it is worth to document it

In order to test if a nested attribute is assignable, `permitted attribute` can also accept a hash, and inside a hash you can insert multiple attributes. Checking the number of parameters sent to the matcher is possible but not trivial, what should we do?

**Comment by tagliala** [comment]
Also, should the description be `mass assignment of`?

**Comment by chrisalley** [comment]
For `permit_attribute`, accepting a single attribute (without children) would be the use case. If checking an attribute with children, we should require that `permit_attributes` is used since we are testing more than one attribute.

**Comment by chrisalley** [comment]
I've created another beta for 3.0. I'd like to get that stable and released first to avoid scope creep, but we can release this soon after as 3.1.

**Comment by tagliala** [comment]
Rebased on main, updated changelog, bumped up version number to 3.1.0

File: spec/pundit/matchers/permit_attributes_matcher_spec.rb [hunk 1]
```diff
@@ -21,14 +21,30 @@
     )
   end
 
-  context 'without actions' do
+  context 'without attributes' do
```
Comment (tagliala) [inline spec/pundit/matchers/permit_attributes_matcher_spec.rb:24]: I think it is not worth to backport this to main

**Comment by chrisalley** [comment]
> Also, should the description be `mass assignment of`?

The code looks good. We should change the Readme section titles to reflect new naming, e.g.

* Testing if an Attribute is Authorised
* Testing if Multiple Attributes are Authorised
* Testing if Attributes are Authorised for Particular Actions

I also think we should demonstrate the `permit_mass_assignment_of` syntax as an alternative at the bottom of each of these three sections, similar to the "The following examples are equivalent" section for `forbid_actions`.

File: README.md [hunk 2]
```diff
@@ -72,6 +72,10 @@ files (by convention, saved in the `spec/policies` directory).
   are permitted by the policy.
 - `permit_edit_and_update_actions` Tests that both the edit and update actions
   are permitted by the policy.
+- `permit_attribute` Tests that an attribute, passed in as a parameter,
+  is permitted by the policy for mass assignment.
+- `permit_attributes` Tests that a list of attributes, passed in as a parameter,
+  are permitted by the policy for mass assignment.
 - `permit_mass_assignment_of(:attribute_name)` or
   `permit_mass_assignment_of(%i[attribute1 attribute2])` Tests that mass
```
Comment (chrisalley) [inline README.md:80]: I think we can change this to simply:

- `permit_mass_assignment_of` Alternative syntax for `permit_attributes`, used for backward compatibility with older test suites.

Same for `forbid_mass_assignment_of`.

That explains the intent of having the additional matcher in a concise sentence.

**Review by chrisalley — APPROVED** [review]

**Pull request merged**